zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -98,20 +126,122 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+            "A hash digest fits in a single field element."
+        );
+        require(
+            length_bit_residual < field_capacity,
+            "Too many input and output notes considered."
+        );
     }
 
     // ============================================================================================ //
     // Reminder: Remember that the primary inputs are ordered as follows:
     // We make sure to have the primary inputs ordered as follow:
-    // [Root, NullifierS, CommitmentS, value_pub_in, value_pub_out]
+    // [Root, NullifierS, CommitmentS, h_sig, h_iS, Residual Field Element(S)]
     // ie, below is the index mapping of the primary input elements on the protoboard:
     // - Index of the "Root" field elements: {0}
     // - Index of the "NullifierS" field elements: [1, NumInputs + 1[
     // - Index of the "CommitmentS" field elements: [NumInputs + 1, NumOutputs + NumInputs + 1[
-    // - Index of the "v_pub_in" field element: {NumOutputs + NumInputs + 1}
-    // - Index of the "v_pub_out" field element: {NumOutputs + NumInputs + 1 + 1}
+    // - Index of the "h_sig" field element: {NumOutputs + NumInputs + 1}
+    // - Index of the "Message Authentication TagS" (h_i) field elements: [NumOutputs + NumInputs + 1 + 1, NumOutputs + NumInputs + 1 + NumOuputs [
+    // - Index of the "Residual Field Element(s)" field elements: [NumOutputs + NumInputs + 1 + NumOuputs + 1 , NumOutputs + NumInputs + 1 + NumOuputs + 1 + nb_field_residual]
+    //
+    // The Residual field elements are structured as follows:
+    // - v_pub_in [0, public_value_length[
+    // - v_pub_out [public_value_length, 2*public_value_length[
+    // - h_sig remaining bits [2*public_value_length, 2*public_value_length + (digest_length-field_capacity)[
+    // - nullifierS remaining bits [2*public_value_length + (digest_length-field_capacity), 2*public_value_length + (1+NumInputs)*(digest_length-field_capacity)[
+    // - commitmentS remaining bits [2*public_value_length + (1+NumInputs)*(digest_length-field_capacity), 2*public_value_length + (1+NumInputs+NumOutputs)*(digest_length-field_capacity)[
+    // - message authentication tagS remaining bits [2*public_value_length + (1+NumInputs+NumOutputs)*(digest_length-field_capacity), 2*public_value_length + (1+2*NumInputs+NumOutputs)*(digest_length-field_capacity)]
     // ============================================================================================ //
 
+    // This function is used to extract the public values (vpub_in and vpub_out) from the residual field element(S)
+    function assemble_public_values(uint[] memory primary_inputs) public pure returns (uint64 vpub_in, uint64 vpub_out){
+        // We know vpub_in corresponds to the first 64 bits of the first residual field element after padding.
+        // We retrieve the public value in, remove any extra bits (due to the padding) and inverse the bit order
+        // N.B. Bytes.int256ToBytes8(x) extract the last 8 bytes of x
```

Comment:
`int256ToBytes8` some function names are not consistent in `Bytes.sol`. This dates back to a long time ago and is not strictly relevant to this PR. I'm just flagging it as a quick thing to fix, in the future, to improve code consistency

Reply:
Note: This can be fixed in another PR as it is not strictly related to this one

Reply:
Function name changed to `extract_last_8_bytes`